numerics/root_finders_test.cpp

```diff
@@ -54,7 +57,134 @@ TEST_F(RootFindersTest, SquareRoots) {
+  // This test constructs a pathological function for which Brent is quadratic.
+  // The construction follows [WG13] and the notation therein.
+  // We will search for a root of the constructed function over [a, b].
+  double const a = 1;
+  double const b = 2;
+
+  // We first construct the sequence of points X at which Brent will be made to
+  // evaluate the function; see section 4.1.
+  std::vector<double> X;
+  // The indices i such that Brent will perform a bisection at X[i].
+  std::set<int> expected_bisections;
+
+  // This tolerance is not the same as the one in our implementation of Brent:
+  // this is because that one is overly close to the machine precision, so that
+  // attempting to perform as many interpolation steps as theoretically possible
+  // before the bisection leads to early bisection in practice, because the
+  // interval shrinkage becomes too great as the convergent values get
+  // discretized.
+  double const δ = 0x1p-50;
+  // The factor p theoretically only needs to be larger than √2, but, for
+  // similar reasons, bringing it closer (e.g., 17/12) requires increasing δ,
+  // which overall reduces the number of evaluations required.
+  double const p = 1.5;
+
+  X.push_back(b);
+  int bisections = 0;
+  for (; a < X.back(); ++bisections) {
+    // k is the number of interpolation steps on the interval [a, X.back()]
+    // before a bisection happens again.
+    int const k = std::round(std::log2((X.back() - a) / δ));
+    for (int j = 1; j <= k; ++j) {
+      X.push_back(X.back() - std::pow(p, k - j) * δ);
+    }
+    expected_bisections.insert(X.size() - 1);
+    X.push_back(a + (X.back() - a) / 2);
   }
+
+  // [WG13], equation (5).
+  auto const s = [](double xa, double xc, double xb, double fa) {
+    return fa * (xb - xc) / (xa - xc);
+  };
+
+  // [WG13], equations (7) and (9).
+  auto const q =
+      [](double xa, double xd, double xc, double xb, double fa, double fb) {
+        double const α = (xa - xd) * fb + (xd - xb) * fa;
+        double const β = (xb - xd) * Pow<2>(fa) + (xd - xa) * Pow<2>(fb);
+        double const γ = fa * fb * (fb - fa) * (xc - xd);
+        return -2 * γ / (β + Sqrt(Pow<2>(β) - 4 * α * γ));
+      };
+
+  // Capture everything but the function itself by copy so we have some
+  // confidence that the function is pure.
+  // We capture f because it is defined recursively.
+  // If non-null, |*evaluations| is incremented when the function is called.
+  // If |expect_brent_calls| is true, this function checks that x is an element
+  // of the expected sequence X.
+  std::function<double(double, int*, bool)> f =
+      [&f, s, q, a, b, X, expected_bisections](
+          double const x,
+          int* const evaluations,
+          bool const expect_brent_calls) -> double {
+    double const f_a = -100;
+    if (evaluations != nullptr) {
+      ++*evaluations;
+    }
+    if (x == a) {
+      return f_a;
+    }
+    if (x == b) {
+      return s(a, X[1], b, f_a);
```

Comment:
Document that your `X` is in reverse order wrt the paper, so `X[1]` is really `X_{m-1}`.  Or reverse the vector after you've constructed it.